File: app/src/main/java/com/kkndesasendang/sendangsmartlearning/ui/learn/selectMaterial/MaterialListAdapter.java
package com.kkndesasendang.sendangsmartlearning.ui.learn.selectMaterial;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.kkndesasendang.sendangsmartlearning.R;
import com.kkndesasendang.sendangsmartlearning.model.MaterialModel;

import java.util.ArrayList;

public class MaterialListAdapter extends RecyclerView.Adapter<MaterialListAdapter.MaterialListViewHolder> {
    private ArrayList<MaterialModel> materialModelList = new ArrayList<>();

    private OnItemClickCallback mOnItemClickCallback;

    interface OnItemClickCallback {
        void onItemClick(MaterialModel data, View view);
    }

    public void setOnItemClickCallback(OnItemClickCallback onItemClickCallback){
        mOnItemClickCallback = onItemClickCallback;
    }

    public void updateDataset(ArrayList<MaterialModel> dataset) {
        materialModelList = dataset;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public MaterialListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_material_list, parent, false);
        return new MaterialListViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MaterialListViewHolder holder, int position) {
        holder.bindTo(materialModelList.get(position));
    }

    @Override
    public int getItemCount() {
        return materialModelList.size();
    }

    public class MaterialListViewHolder extends RecyclerView.ViewHolder {
        private TextView tvMaterialList;

        public MaterialListViewHolder(@NonNull View itemView) {
            super(itemView);
            tvMaterialList = itemView.findViewById(R.id.itemMaterialName);
        }

        public void bindTo(final MaterialModel materialModel) {
            tvMaterialList.setText(materialModel.getMaterialName());

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    mOnItemClickCallback.onItemClick(materialModel, view);
                }
            });
        }
    }
}

File: app/src/main/java/com/kkndesasendang/sendangsmartlearning/ui/rvAdapters/MaterialListAdapter.kt
package com.kkndesasendang.sendangsmartlearning.ui.rvAdapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.kkndesasendang.sendangsmartlearning.R
import com.kkndesasendang.sendangsmartlearning.model.GamaselaMaterialModel
import kotlinx.android.synthetic.main.item_generic_list.view.*

class MaterialListAdapter(): RecyclerView.Adapter<MaterialListAdapter.MaterialListViewHolder>() {
    private var topics: Array<out GamaselaMaterialModel> = emptyArray()

    private lateinit var onItemClickCallback: OnItemClickCallback

    interface OnItemClickCallback {
        fun onItemClick(topic: GamaselaMaterialModel)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MaterialListAdapter.MaterialListViewHolder {
        val view: View = LayoutInflater.from(parent.context).inflate(R.layout.item_generic_list, parent, false)
        return MaterialListViewHolder(view)
    }

    override fun onBindViewHolder(holder: MaterialListAdapter.MaterialListViewHolder, position: Int) {
        holder.bind(topics[position])
    }

    override fun getItemCount(): Int {
        return topics.size
    }

    fun setOnItemClickCallback(onItemClickCallback: OnItemClickCallback) {
        this.onItemClickCallback = onItemClickCallback
    }

    fun updateDataset(topics: Array<out GamaselaMaterialModel>) {
        this.topics = topics
        notifyDataSetChanged()
    }

    inner class MaterialListViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(item: GamaselaMaterialModel) {
            itemView.listItemText.text = item.materialName
            itemView.setOnClickListener { onItemClickCallback.onItemClick(item)}
        }
    }
}